skimage/morphology/_flood_fill.py

```diff
@@ -0,0 +1,290 @@
+        initial value of `image` at `seed_point`.  This is fastest.  If a value
+        is given, a comparison will be done at every point and this tolerance
+        on each side of the initial value will also be filled (inclusive).
+    inplace : bool, optional
+        If True, flood filling is applied to `image` inplace.  If False, the
+        flood filled result is returned without modifying the input `image`
+        (default).
+
+    Returns
+    -------
+    filled : ndarray
+        An array with the same shape as `image` is returned with values equal
+        to (or within tolerance of) the seed point set to `new_value`.
+
+    Notes
+    -----
+    The conceptual analogy of this operation is the 'paint bucket' tool in many
+    raster graphics programs.
+
+    Examples
+    --------
+    >>> from skimage.morphology import flood_fill
+    >>> image = np.zeros((4, 7), dtype=int)
+    >>> image[1:3, 1:3] = 1
+    >>> image[3, 0] = 1
+    >>> image[1:3, 4:6] = 2
+    >>> image[3, 6] = 3
+    >>> image
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [1, 0, 0, 0, 0, 0, 3]])
+
+    Fill connected ones with 5, with full connectivity (diagonals included):
+
+    >>> flood_fill(image, (1, 1), 5)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [5, 0, 0, 0, 0, 0, 3]])
+
+    Fill connected ones with 5, with only cardinal direction connectivity:
+
+    >>> flood_fill(image, (1, 1), 5, connectivity=1)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [1, 0, 0, 0, 0, 0, 3]])
+
+    Fill with a tolerance:
+
+    >>> flood_fill(image, (0, 0), 5, tolerance=1)
+    array([[5, 5, 5, 5, 5, 5, 5],
+           [5, 5, 5, 5, 2, 2, 5],
+           [5, 5, 5, 5, 2, 2, 5],
+           [5, 5, 5, 5, 5, 5, 3]])
+    """
+    # Correct start point in ravelled image - only copy if non-contiguous
+    image = np.asarray(image)
+    copied = False
+
+    if not image.flags.contiguous:
+        if inplace:
+            warnings.warn('Non-contiguous array passed as `image`; this will '
+                          'be converted to a contiguous array as a copy.')
+            # Flood fill requires some sort of contiguity - this makes a copy
+            image = np.ascontiguousarray(image)
+            copied = True
+
+    if not inplace:
+        if not copied:
+            output = image.copy()
+        else:
+            output = image
+        output[flood(image, seed_point, selem=selem, connectivity=connectivity,
+                     tolerance=tolerance)] = new_value
+        return output
+    else:  # inplace
+        image[flood(image, seed_point, selem=selem, connectivity=connectivity,
+                    tolerance=tolerance)] = new_value
+        return image
+
+
+def flood(image, seed_point, *, selem=None, connectivity=None, tolerance=None):
```

Comment:
Has there been a discussion to add a parameter `indices` that will change the returned output similar to the behavior in `local_maxima`? 

https://github.com/scikit-image/scikit-image/blob/39176efc2ad4e1131b57bb14f42bca335d43e399/skimage/morphology/extrema.py#L356-L360

Might be useful for the user and make the API more consistent. Using `np.nonzero` it's really trivial to add.

Reply:
@lagru we had a discussion about this that GitHub has probably helpfully masked. =P Also some of it may have happened offline while at SciPyConf — sorry if we did an incomplete job transcribing to here.

The basic idea behind not going this route is:

- we want to return the `bool` array because it is produced as a side-effect of the running of the function, so calling `np.nonzero` on it is extra overhead that users may not need.
- we generally want to avoid changing return types based on the value of some inputs. Even though it is Pythonic to do this sort of thing, it makes it harder (sometimes impossible) to statically check the type correctness of code.

Even though we've got some examples where we do this, the question is how we want to proceed going forward.

Finally, since this is a trivial change to add later without breaking the existing API, I suggest that this should not hold up this PR.

Reply:
@jni Okay, good to know. Sounds sensible. Perhaps we should transcribe that somewhere? Maybe the developer guide?